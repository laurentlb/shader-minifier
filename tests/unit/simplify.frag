#version 330

int i;

float bar(float x)
{
	float a = x;
	a = 6.;
	a = a + x;
	
	float b = x;
	b = b + x;
	b = 34.;
	b = b + x;

    float arr[2] = float[2](7.,8.);
	arr = float[2](5.,float(arr.length()));
	
	float m = 14.+length(vec3(sin(a *= 10), sin(a *= 20), sin(a *= 30)))+b++;
	m = arr[1] + arr[i++];
		
	return a + b + m;
}
float baz(float a)
{
	float b = a + 4.;
	b += sin(a);
	float c = b + 5.;
	c += sin(b);
	return c + (-(c - -c));
}

int n = 2;
float y = 47.;
out vec3 outputvar;

float foo(float a) { if (n == 1) return 0.; if (n == 2) return a; return 1.; }
float foo(float a, float b) { if (y > a) return 0.; if (y < b) return a; return 1.; }

void notMain(float x) {
	vec3 v = vec3(foo(42.) + foo(50., 70.));
	outputvar.rgb = v + vec3(bar(x) + baz(x));
}

// #442: function inlining when a parameter is read more than once, but the argument expr is pure and trivial, like a variable that's read (but not written)
bool IsMoto(float mid) {
  return mid>=1.&&mid<=8.;
}
int foo442(float x) {
  if (IsMoto(x)) return 1;
  return 2;
}

// #450: argument inlining of read-only globals
in vec3 camPos;
float len(vec3 pos) {
  return length(pos);
}
float main450() {
  float a = len(camPos);
  float b = len(camPos);
  return a + b;
}
