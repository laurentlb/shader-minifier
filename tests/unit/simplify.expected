#version 330

int i;
float bar(float x)
{
  float a=6.+x,b=34.+x,arr[2]=float[2](5.,float(float[2](7.,8.).length()));
  x=(a*=10,a*=20,a*=30,b++,arr[1]+arr[i++]);
  return a+b+x;
}
float baz(float a)
{
  a=a+4.+sin(a);
  a=a+5.+sin(a);
  return a-a-a;
}
out vec3 outputvar;
void notMain(float x)
{
  outputvar.xyz=vec3(92)+vec3(bar(x)+baz(x));
}
int foo442(float x)
{
  return x>=1.&&x<=8.?
    1:
    2;
}
in vec3 camPos;
float len()
{
  return length(camPos);
}
float main450()
{
  return len()+len();
}
